### PR TITLE
feat: add shared/pkg/bucketing canonical library

### PR DESCRIPTION
**Pull request by bjcoombs**
## Summary

- Add `shared/pkg/bucketing` package providing canonical bucket ID calculation to prevent Bucket Drift across services
- `CalculateBucketID()` produces deterministic IDs in format `dimension_instrument_attr1=val1_attr2=val2` with sorted attributes
- `GetDimension()` maps instrument codes to their dimension (CURRENCY, ENERGY, COMPUTE, CARBON, DATA, VOLUME, MASS, COUNT) with thread-safe `RegisterDimension()` for custom instruments
- `ParseBucketID()` / `ValidateBucketID()` for decomposition and validation with round-trip fidelity
- `scripts/lint-bucket-id.sh` CI lint script detects manual bucket_id string construction outside the canonical library

Task Master: valuation-engine.8

## Test plan

- [x] Unit tests for CalculateBucketID with monetary, energy, compute, carbon instruments
- [x] Deterministic key generation verified across 1000 calls (drift prevention)
- [x] Attribute sort order independence verified
- [x] Edge cases: nil/empty attributes, empty instrument code, empty dimension
- [x] GetDimension mapping for all ISO 4217 currencies and asset types
- [x] RegisterDimension for custom instrument codes
- [x] ParseBucketID round-trip fidelity
- [x] ValidateBucketID format validation
- [x] Lint script passes on current codebase (no false positives)

### CODE REVIEW

File: shared/pkg/bucketing/bucketing.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+		// Carbon -> CARBON
+		"CARBON_CREDIT": "CARBON",
+		"CARBON_TONNE":  "CARBON",
+
+		// Volume -> VOLUME
+		"WATER_LITRE": "VOLUME", //nolint:misspell // British spelling matches domain convention
+
+		// Mass -> MASS
+		"CARBON_TONNES": "CARBON",
+	}
```
[the remark this answers is not in the document]
Comment (bjcoombs) [reply shared/pkg/bucketing/bucketing.go:77]: Fixed. Removed the duplicate entry under 'Mass -> MASS' comment. CARBON_TONNES correctly maps to CARBON (it measures carbon emissions, not generic mass). Added test coverage for GetDimension("CARBON_TONNES").

File: shared/pkg/bucketing/bucketing.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+			}
+			attrs[seg[:eqIdx]] = seg[eqIdx+1:]
+		}
+	}
+
+	return BucketIDParts{
+		Dimension:      dimension,
+		InstrumentCode: instrumentCode,
+		Attributes:     attrs,
+	}, nil
```
[the remark this answers is not in the document]
Comment (bjcoombs) [reply shared/pkg/bucketing/bucketing.go:221]: Fixed all three issues:
1. ParseBucketID now rejects empty dimension/instrument segments (returns ErrInvalidFormat)
2. ParseBucketID now rejects empty attribute keys (eqIdx <= 0)
3. CalculateBucketID now panics on attribute keys/values containing underscores or '=' (since these produce unparseable bucket IDs)

Added test coverage for all edge cases.
